src/frontend/src/binder/select.rs

```diff
@@ -153,35 +192,185 @@ impl Binder {
-        let bound_columns = columns
-            .iter()
-            .filter_map(|column| {
-                if !column.is_hidden {
-                    Some((
-                        InputRef::new(column.index, column.data_type.clone()).into(),
-                        Some(column.column_name.clone()),
-                    ))
+        let (binding, idents) = self.extract_binding_and_idents(expr, ids.to_vec())?;
+        let (exprs, column) = Self::bind_field(
+            InputRef::new(binding.index, binding.desc.data_type.clone()).into(),
+            &idents,
+            binding.desc.clone(),
+            true,
+        )?;
+        Ok((
+            exprs,
+            column
+                .field_descs
+                .iter()
+                .map(|f| Some(f.name.clone()))
+                .collect_vec(),
+        ))
+    }
+
+    /// Bind single field column, e.g. `(table.v1).v2`.
+    /// Will return Field type `FunctionCall` and alias.
+    pub fn bind_single_field_column(
+        &mut self,
+        expr: Expr,
+        ids: &[Ident],
+    ) -> Result<(ExprImpl, Option<String>)> {
+        let (binding, idents) = self.extract_binding_and_idents(expr, ids.to_vec())?;
+        let (exprs, column) = Self::bind_field(
+            InputRef::new(binding.index, binding.desc.data_type.clone()).into(),
+            &idents,
+            binding.desc.clone(),
+            false,
+        )?;
+        Ok((exprs[0].clone(), Some(column.name)))
+    }
+
+    /// Bind field in recursive way, each field in the binding path will store as `literal` in
+    /// exprs and return.
+    /// `col` describes what `expr` contains.
+    pub fn bind_field(
+        expr: ExprImpl,
+        idents: &[Ident],
+        desc: ColumnDesc,
+        wildcard: bool,
+    ) -> Result<(Vec<ExprImpl>, ColumnDesc)> {
+        match idents.get(0) {
+            Some(ident) => {
+                let (field, field_index) = desc.field(&ident.value)?;
+                let expr = FunctionCall::new_with_return_type(
+                    ExprType::Field,
+                    vec![
+                        expr,
+                        Literal::new(Some(field_index.to_scalar_value()), DataType::Int32).into(),
+                    ],
+                    field.data_type.clone(),
+                )
+                .into();
+                Self::bind_field(expr, &idents[1..], field, wildcard)
+            }
+            None => {
+                if wildcard {
+                    Self::bind_wildcard_field(expr, desc)
                 } else {
-                    None
+                    Ok((vec![expr], desc))
                 }
+            }
+        }
+    }
+
+    /// Will fail if it's an atomic value.
+    /// Rewrite (expr:Struct).* to [Field(expr, 0), Field(expr, 1), ... Field(expr, n)].
+    pub fn bind_wildcard_field(
+        expr: ExprImpl,
+        desc: ColumnDesc,
+    ) -> Result<(Vec<ExprImpl>, ColumnDesc)> {
+        if desc.field_descs.is_empty() {
```

Comment:
This is not always correct. Assume an empty schema:

```
message Foo {
}
```

Will postgres fail on `foo.*`?